saleor/product/models.py

```diff
@@ -180,67 +191,71 @@ def sort_by_attribute(self, attribute_pk: Union[int, str], ascending: bool = Tru
+                concatenated_values=Case(
+                    # If the product has no association data but has
+                    # the given attribute associated to its product type,
+                    # then consider the concatenated values as empty (non-null).
+                    When(
+                        Q(product_type_id__in=product_types_associated_to_attribute)
+                        & Q(filtered_attribute=None),
+                        then=models.Value(""),
                     ),
+                    default=StringAgg(
+                        F("filtered_attribute__values__name"),
+                        delimiter=",",
+                        ordering=(
+                            [
+                                f"filtered_attribute__values__{field_name}"
+                                for field_name in AttributeValue._meta.ordering or []
+                            ]
+                        ),
+                    ),
+                    output_field=models.CharField(),
                 ),
-                output_field=models.CharField(),
-            ),
-            concatenated_values_order=Case(
-                # Make the products having no such attribute be last in the sorting
-                When(concatenated_values=None, then=2),
-                # Put the products having an empty attribute value at the bottom of
-                # the other products.
-                When(concatenated_values="", then=1),
-                # Put the products having an attribute value to be always at the top
-                default=0,
-                output_field=models.IntegerField(),
-            ),
-        )
+                concatenated_values_order=Case(
+                    # Make the products having no such attribute be last in the sorting
+                    When(concatenated_values=None, then=2),
+                    # Put the products having an empty attribute value at the bottom of
+                    # the other products.
+                    When(concatenated_values="", then=1),
+                    # Put the products having an attribute value to be always at the top
+                    default=0,
+                    output_field=models.IntegerField(),
+                ),
+            )
 
         # Sort by concatenated_values_order then
         # Sort each group of products (0, 1, 2, ...) per attribute values
         # Sort each group of products by name,
         # if they have the same values or not values
-        qs = qs.order_by("concatenated_values_order", "concatenated_values", "name")
-
-        # Descending sorting
-        if not ascending:
-            return qs.reverse()
-        return qs
+        ordering = "-" if descending else ""
+        return qs.order_by(
+            f"{ordering}concatenated_values_order",
+            f"{ordering}concatenated_values",
+            f"{ordering}name",
+        )
```

Comment:
Did you check the cost of the queries? Those were very optimized, and a lot of work was put into making them as optimal as possible.

Could you test them and see how they perform underneath? Using `EXPLAIN`

Reply:
I didn't changing anything here, except that I moved `if` statement from resolver to the beginning of this method

and this I don't think changing sorting direction make this code less optimal